Pagination Implementation

razzlero opened the pull request:
A Pull request for implementing  #36 

I have coded up a potential implementation of Pagination.
The parameters Pagenum, PageAfter, and PageSize are now accepted as query string parameters and are processed.
Pageafter is given a higher priority than pagenum, so if they are used in the same query string pageAfter is the one which will be used.

The changes were primarily made to the createMap function inside lib/operations/collection/get.js .
The createList function has been modified to first run createMap,and then generate the list form that map. This added a bit of extra processing, but it makes sure the 2 functions return the same results.

I decided to not sort the results before performing pagination. I thought it would be best to allow users to have control over the order of their results. 
Currently the results should be predictable enough in most cases for pagination to work without the need to sort first. 
In future it might be interesting to add an optional sort parameter though.

Also in future it might be a good idea to implement a way of differentiating the query string parameters passed through for searching from the special query string parameters. As it works now if by some slim chance a user has a property called "pageNum", "pageAfter", or "pageSize", and they want to search against that parameter, they won't be able to because of these special functions taking priority.

ronkorving commented on lib/operations/collection/get.js:
```diff
@@ -1,39 +1,79 @@
+	if (obj && obj.hasOwnProperty(prop)) {
+		retval = obj[prop];
+		delete obj[prop];
+		// If the property is a numeber stored as a string we convert it to an actual number
```
typo: number

ronkorving commented:
If arguments passed are invalid (say a page size of 0), I would rather see an error than get the entire collection back, which I hope you agree is invalid behavior.

razzlero commented on lib/operations/collection/get.js:
```diff
@@ -1,39 +1,79 @@
+		retval = obj[prop];
+		delete obj[prop];
+		// If the property is a numeber stored as a string we convert it to an actual number
+		if (!isNaN(retval)) {
```
isNaN can be a bit confusing, especially needing to use it as a double negative like this can be quite confusing. It'll think up an alternative and replace it.

razzlero commented:
I'm actually not sure why I decided to return the full collection instead of an error. I had some train of thought which lead to it, but returning a error would be better. I'll look into adding an error.

I had a couple of other solutions which would have avoided looping the list twice, but I thought they would be a bit messy. I decided to choose readability/maintainability over performance in this case. 

Another idea I had was:

1. Create a generic createResult function bases on the createList/createMap functions.
2. Add a parameter to the function specifying the output type, this would be either 'List' or 'map'.
3. Set the result type to be either a list or an object depending on the parameter passed.
4. The assignment loop have an if statement checking the output type and assign it as a map or list as specified.

Now that I think about it again, maybe that approach wouldn't be too bad.

razzlero commented:
The most recent commits add errors when pageination numbers are too small or are invalid numbers. 
I also throw a 404 error if they try to pageAfter using an id which does not exist.
I decided to consider page numbers which are too big as valid (and just return no results with no error).

I extracted the pagination logic mostly into its own function. (All the ifs in that function make it pretty messy to look at though).

If I get a chance I'll try coding up the generic function which can act as either createMap or createList.

razzlero commented on lib/operations/collection/get.js:
```diff
@@ -4,51 +4,73 @@ const PAGE_AFTER = 'pageAfter';
 	// Get special query string properties and remove them from the query object.
+	const pageSize = takeProperty(query, PAGE_SIZE);
 	const pageAfter = takeProperty(query, PAGE_AFTER);
 	const pageNum = takeProperty(query, PAGE_NUM);
```
I could have calculated pageSize, pageAfter, and pageNum inside the getPagination function, but I decided to get them here. This is because I wanted to make it easier to see that these properties are being removed from the 'query' object before it is accessed in the loop.

razzlero commented:
I combined createList and createMap into a single function.
I also added a slight optimisation. When createMap is being performed with no query string no processing is required, so in that case the existing map is returned without doing any processing.

ronkorving commented:
I agree with your statement about all the ifs making it rather messy. My own suggestion would have been to add 2 arguments to the function, so the function could fill both a map and an array if it had to. That is, rather then creating the result values *inside* the function, pass them in. You would need only 2 if-statements and no else-blocks.

razzlero commented:
Looking at the other places where setStatus is used I think I see what you were getting at. I think it's more that I didn't follow the application's architecture enough while using it.

I see in most other places the status is being set just before returning the cb. So I think you expected me to do something similar?
Now that I i see that, I realise it is a better way of doing it. You would be able to clearly see where the response is set rather than in being hidden inside a function or another call.
(Maybe the way I wrote it gave the impression that I thought it was instantly throwing an error at that point?)

Just to make I understand you sure before I make any code changes. Is that what you were talking about?
If I'm wrong feel free to correct me.

ronkorving commented:
My point about setStatus was that calling that, doesn't automagically prevent the body payload from being generated, but in the case of these errors we don't want a payload. Doing it that early on makes it harder to choose to send or not send a payload. So you're on the right track, but it's not because of convention, but rather because the current implementation in your PR does too much work in the case of an error.

razzlero commented:
OK, I see what you mean now.
I have a couple of ideas how to go about fixing it up. I didn't get much time to work on it today, but tomorrow I should be able to make the changes.

razzlero commented:
A lot has changed in this version.
Firstly I changed "getResult" to "getresults" which as you suggested will generate both a list and a map.

Secondly I have extracted the validation out of createResults/getPagination and perform it in a separate function called validateProperties. I do this validation near the beginning of module.exports to make sure errors are caught early and to make sure data is in the correct format to work smoothly afterwards.
I tried to keep the validation readable by avoiding too many nested if statements and storing results of boolean expressions in consts.

I didn't really like that I was modifying the query object, so I now copy all the properties out of the query object into  a 'properties' object including the properties for performing matching.

ronkorving commented:
Thanks for the updates!

When I mentioned the results list and map objects, I really meant passing them into your function empty, and have your function fill them if they are set. Instead of what you have now, which is filling up both when you only need one.

I like what you did with the properties object, making a clear distinction between what we're matching on and how we paginate.

One more hint to make your code more compact/readable (IMHO) is that you don't need to do a `typeof` check to test if an object's property value is `undefined`. Also, are you familiar with `Number.isInteger`?

razzlero commented:
I thought it seemed strange to be filling an object we wouldn't be using. I should have reread your comment when I started coding that feature.

The `typeof` call was a remnant from before I stored those variables in the properties object. Good job noticing it.

Funnily, I thought at some stage I triggered an issue with `Number.isInteger` in lint or JSCS, so I avoided using it. But I see now it works fine. My memory must be playing tricks on me.

I have changed the getResult Function to take an empty map and/or list as a parameter and fills them if they are set.
I have replaced the `typeof` calls with `properties.hasOwnProperty`.
And I have removed the `isNumeric` function and replaced it with calls to `Number.isInteger`

ronkorving commented on lib/operations/collection/get.js:
```diff
@@ -1,46 +1,115 @@
+	let status = 200;
+
+	const hasPageSize = properties.hasOwnProperty('pageSize');
+	const pageSizeIsNumber = Number.isInteger(properties.pageSize) && typeof (properties.pageSize) === 'number';
```
If something is `Number.isInteger`, you won't have to do a typeof check anymore :) If it's an integer, it's a number.

razzlero replied:
Another bit I missed by being too shortsighted with my changes. I'll have to pay more attention to full lines in future.

ronkorving commented:
Please review your use of HTTP status codes. I think you're not applying them correctly in the error cases. Feedback welcome of course :)

razzlero commented:
I have set all the HTTP status codes back to 400. I was trying to be more specific, but after some googling I see 400 being used in similar situations.
If you think using 400 is wrong let me know.

ronkorving commented on lib/operations/collection/get.js:
```diff
@@ -26,28 +26,28 @@ function validateProperties(map, properties) {
+			status = 400; // Bad Request: Invalid number
 		} else if (hasPageAfter && !pageAfterKeyFound) {
-			status = 404; // Not found
+			status = 400; // Bad Request: PageAfter not found
```
Is that a bad request, a 404, or an empty result?
My gut tells me 404. Thoughts welcome.

razzlero replied:
I think you're right, 404 is probably most correct in this case. 
The most likely scenario of it occurring might be if somebody tries to access a resource which has been removed which would be a 404.
I have changed it back to 404 now.

ronkorving commented:
I think it's right in most cases. I've noted the one exception I think exists.

ronkorving commented:
So now that the feature is basically complete, I think there's one bit missing still. When paginating an a key/value map (instead of an array), the order is not predetermined. Perhaps we can sort the keys before paginating (at least when we ask for map-output)?

razzlero commented:
I've added an orderBy function which allows you to declare a property to order by.

I currently haven't set any default value orderBy.
I'm a little unsure if there should be a default orderBy or not. If we do go with a default I would use the id as the default value.
I'm just thinking that if there is no default then the developers using the tool can pass data into the system in a particular order if they wish.

What's your opinion? Do you think it would be better to add a default, or have no default?
Or is there already some processing which I am not aware of which could interfere with the order of the data passed in by the developers?

ronkorving commented:
The most important part for me is that pagination *works*. The problem with `Object.keys(map)` is that it does not guarantee the order to be the same every time you call it. Currently, the V8 engine does seem to return the keys in the same order each time, but JavaScript (the language) does not guarantee this. Therefore, I would be less worried if the keys would be sorted (in *any* way at all, as long as it's deterministic).

razzlero commented:
After doing my research I now realise I was mislead by something I heard before. I heard before that ES6 implemented maps with consistent ordering. I now realise that a new "map" datatype was introduced. I thought they were talking about the existing objects.
I have added a default sort which just does a keys.sort();

Seeing as I have that sorting added I might as well leave it in place for ow. Maybe in a future pull request it can be expanded on for more advanced behaviour.

ronkorving commented:
Works for me, thanks 👍

ronkorving commented:
Thanks again :+1:
